### PR TITLE
Optimize bytecode amount

### PR DESCRIPTION
**Pull request by luontola**
Fixes https://github.com/weavejester/hiccup/issues/205

This PR does two optimizations which reduce the amount of generated bytecode:
1. Optimize `hiccup.compiler/build-string` to concatenate consecutive string literals at compile time
2. Optimize `hiccup.compiler/compile-element ::literal-tag` to generate only once the code for an element's children and ending tag

Example of the generated code:
- [input](https://gist.githubusercontent.com/luontola/bb95eb7779c3d18b5486cdfe360f843c/raw/5269cdc68e000079ddae0be8194762ace922049f/example.clj)
- [output before](https://gist.githubusercontent.com/luontola/bb95eb7779c3d18b5486cdfe360f843c/raw/5269cdc68e000079ddae0be8194762ace922049f/output-12345.clj)
- [output after](https://gist.githubusercontent.com/luontola/bb95eb7779c3d18b5486cdfe360f843c/raw/5269cdc68e000079ddae0be8194762ace922049f/output-12345-fixed.clj)

I did a little benchmarking and this improved the performance of my code by 7-14%. This improvement was due to concatenating the strings at compile time. It could be optimized further by using only one java.lang.StringBuilder, avoiding the multiple intermediate StringBuilder->String->StringBuilder conversions. I'll have a look at that later in another PR.

### CODE REVIEW

File: test/hiccup2/core_test.clj [hunk 1]
```diff
@@ -212,3 +212,50 @@
...
+    (is (= (str (html {:mode :xhtml} [:br (identity nil)])) "<br></br>"))
+    (is (= (str (html {:mode :html} [:br (identity nil)])) "<br></br>"))
+    (is (= (str (html {:mode :xml} [:br (identity nil)])) "<br></br>"))
+    (is (= (str (html {:mode :sgml} [:br (identity nil)])) "<br></br>")))
+  (testing "compile-time both absent: nil child"
+    ;; TODO: same as above, but more of the user's fault to write a nil literal inside a void element
+    (is (= (str (html {:mode :xhtml} [:br nil])) "<br></br>"))
+    (is (= (str (html {:mode :html} [:br nil])) "<br></br>"))
+    (is (= (str (html {:mode :xml} [:br nil])) "<br></br>"))
+    (is (= (str (html {:mode :sgml} [:br nil])) "<br></br>")))
```
[the remark this answers is not in the document]
Comment (luontola) [reply test/hiccup2/core_test.clj:239]: If you think this is a bug that should be fixed, I volunteer to work on it.

File: test/hiccup/compiler_test.clj [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+            [hiccup.compiler :as compiler]))
+
+(deftest test-concatenate-strings
+  (let [concatenate-strings #'compiler/concatenate-strings]
+    (testing "empty collection"
+      (is (= [] (concatenate-strings nil)))
+      (is (= [] (concatenate-strings []))))
+    (testing "concatenates consecutive strings, ignores other types"
+      (is (= ["abc"] (concatenate-strings ["a" "b" "c"])))
+      (is (= [1 "ab" 2 "cd" 3] (concatenate-strings [1 "a" "b" 2 "c" "d" 3]))))))
```
Comment (weavejester) [inline test/hiccup/compiler_test.clj:12]: We shouldn't be testing private functions on their own, as it should be possible to refactor the internals of a namespace without causing the tests to fail.

Can we perform the same effective test in `hiccup.core_test`, using `html`?

**Review by weavejester — COMMENTED** [review]
Thank you very much for your work on this, I have some comments/changes needed before it can be merged.

File: test/hiccup/compiler_test.clj (hunk 2 above)
Comment (luontola) [reply test/hiccup/compiler_test.clj:12]: It's only used by the macro, so an integration test would not be able check whether `concatenate-strings` works or whether it's just an alias for `identity`.

Comment (weavejester) [reply test/hiccup/compiler_test.clj:12]: But we can check the generated code with `macroexpand-1`, right?

File: test/hiccup2/core_test.clj (hunk 1 above)
[the remark this answers is not in the document]
Comment (luontola) [reply test/hiccup2/core_test.clj:239]: I'll change this test when I write the fix. For now it's useful for it to document the current behavior, and it was also critical in preventing my optimizations from changing the current behavior. It covers edge cases that the existing tests didn't cover, and it saved me a already couple of times during my refactoring.

File: src/hiccup/compiler.clj [hunk 3]
```diff
@@ -314,22 +333,24 @@
 (defmethod compile-element ::literal-tag
   [[tag attrs & content]]
   (let [[tag tag-attrs _] (normalize-element-form [tag])
-        attrs-sym         (gensym "attrs")]
+        attrs-sym (gensym "attrs")]
```
[the remark this answers is not in the document]
Comment (luontola) [reply src/hiccup/compiler.clj:336]: What code formatter do you use? I wasn't able to configure Cursive to produce exactly the same style.

File: test/hiccup/compiler_test.clj (hunk 2 above)
Comment (luontola) [reply test/hiccup/compiler_test.clj:12]: Checking the outcome of macroexpand would produce extremely fragile tests. It would hinder refactoring instead of helping it, because it would cause refactoring to break tests.

File: src/hiccup/compiler.clj (hunk 3 above)
[the remark this answers is not in the document]
Comment (weavejester) [reply src/hiccup/compiler.clj:336]: I'm using Emacs/CIDER. Does Cursive try to revert aligned let forms?

File: src/hiccup/compiler.clj [hunk 4]
```diff
@@ -314,22 +333,24 @@
...
-                           ~(str "</" tag ">"))
-            `(build-string ~(str "<" tag)
-                           (render-attr-map (merge ~tag-attrs ~attrs-sym))
-                           ~(end-tag)))
-         ~(if (container-tag? tag attrs)
-            `(build-string ~(str "<" tag (render-attr-map tag-attrs) ">")
-                           ~@(compile-seq (cons attrs-sym content))
-                           ~(str "</" tag ">"))
-            (build-string "<" tag (render-attr-map tag-attrs) (end-tag)))))))
+       (build-string
```
[the remark this answers is not in the document]
Comment (weavejester) [reply src/hiccup/compiler.clj:338]: Ah, I see. I'd missed that the `if` and `when` forms are concatenated.

Could you change the indentation so that it's:

```clojure
(build-string
 (if (map? ~attrs-sym)
   ...)
 (when ...))
```

So indented one space, rather than two. Hiccup follows the Clojure Style Guide for it's indentation, and for whatever reason Cursive differs from it.

```clojure
;; Style guide on how to indent lists by default
(foo
 bar)
;; Cursive on how to indent lists by default
(foo
  bar)
```

File: src/hiccup/compiler.clj [hunk 5]
```diff
@@ -314,22 +333,24 @@
...
+                             (render-attr-map (merge ~tag-attrs ~attrs-sym))
+                             ">")
+              `(build-string ~(str "<" tag)
+                             (render-attr-map (merge ~tag-attrs ~attrs-sym))
+                             ~(end-tag)))
+           (build-string ~(str "<" tag (render-attr-map tag-attrs) ">")
+                         ~@(compile-seq [attrs-sym])))
+         ;; content and ending tag (when the above code did not emit an ending tag)
+         (when (or ~(boolean (container-tag? tag content))
+                   (not (map? ~attrs-sym)))
```
Comment (weavejester) [inline src/hiccup/compiler.clj:351]: Could this be written as:

```clojure
~(if (container-tag? tag content)
   `(build-string ~@(compile-seq content) ~(str "</" tag ">"))
   `(when (not (map? ~attrs-sym))
      (build-string ~@(compile-seq content) ~(str "</" tag ">"))))
```

Unless I've misread the code, this should do the same thing, but potentially shave off another form in some circumstances.

File: src/hiccup/compiler.clj (hunk 3 above)
[the remark this answers is not in the document]
Comment (luontola) [reply src/hiccup/compiler.clj:336]: Cursive aligns let forms, but ignores destructurings when deciding how deep to align them.

You could configure a Leiningen alias for formatting the whole project. That would end all discussion about code formatting and PRs could focus on more important things. For example add the [zprint plugin](https://github.com/kkinnear/zprint/blob/main/doc/using/files.md#5-lein-zprint) and configure it to use the [community style](https://github.com/kkinnear/zprint/blob/main/doc/reference.md#community).

File: test/hiccup/compiler_test.clj (hunk 2 above)
Comment (weavejester) [reply test/hiccup/compiler_test.clj:12]: We could do it that way, but it seems simpler just to count the forms and use that as an approximation for the size of the generated bytecode.

To put it another way, it's unlikely that the bytecode is going to significantly increase in size if the number of forms remains the same. The point of tests is to alert us of future regressions, after all. We just need a sufficiently good canary in our coalmine; it doesn't need to be perfect.

Comment (luontola) [reply test/hiccup/compiler_test.clj:12]: Yes, that seems like a good compromise. Counting forms is easier.

Comment (luontola) [reply test/hiccup/compiler_test.clj:12]: I've now written such tests.

Comment (weavejester) [reply test/hiccup/compiler_test.clj:12]: Thanks! The new tests look good. Can you remove this compiler test and namespace from the PR?

File: test/hiccup2/optimizations_test.clj [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            [hiccup2.core :as h]))
+
+(defn- count-forms [data]
+  (let [*counter (atom 0)]
+    (walk/postwalk (fn [x]
+                     (when (seq? x)
+                       (swap! *counter inc))
+                     x)
+                   data)
+    @*counter))
```
Comment (weavejester) [inline test/hiccup2/optimizations_test.clj:14]: What about:

```clojure
(defn- count-forms [data]
  (count (filter seq? (tree-seq coll? seq data))))
```

I believe it'll generate the same result, just a little more functionally.

File: test/hiccup2/optimizations_test.clj [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+(deftest test-count-forms
+  (is (= 1 (count-forms '())))
+  (is (= 3 (count-forms '(if (number? x) (inc x) x)))))
```
Comment (weavejester) [inline test/hiccup2/optimizations_test.clj:18]: I don't think we need to regression test our regression tests, so these can be omitted.

File: test/hiccup2/optimizations_test.clj [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                        `(h/html [:div
+                                  [:p]
+                                  (identity nil)
+                                  [:p]]))
+          pathological (walk/macroexpand-all
+                        `(h/html [:div
+                                  [:p] [:p] [:p] [:p] [:p]
+                                  (identity nil)
+                                  [:p] [:p] [:p] [:p] [:p]]))]
+      #_(pp/pprint pathological)
```
Comment (weavejester) [inline test/hiccup2/optimizations_test.clj:37]: Let's remove debug code from the final commit.

File: test/hiccup2/optimizations_test.clj [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                     example-1
+                     example-2
+                     example-3
+                     example-4
+                     example-5]
+          diffs     (->> examples
+                         (map count-forms)
+                         (partition 2 1)
+                         (map (fn [[a b]] (- b a))))]
+      #_(pp/pprint example-3)
```
Comment (weavejester) [inline test/hiccup2/optimizations_test.clj:88]: Similarly, let's remove this debug code.

File: test/hiccup2/optimizations_test.clj [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                        `(h/html [:div
+                                  [:p] [:p] [:p] [:p] [:p]
+                                  (identity nil)
+                                  [:p] [:p] [:p] [:p] [:p]]))]
+      #_(pp/pprint pathological)
+      (is (= (count-forms baseline)
+             (count-forms pathological)))))
+
+  (testing "code size should grow O(n), instead of O(n^2), as more dynamic first-child elements are added"
+    (let [example-0 (walk/macroexpand-all
```
Comment (weavejester) [inline test/hiccup2/optimizations_test.clj:42]: I like this approach. It's more robust than simply checking for the number of forms against a constant.

**Comment by luontola** [comment]
Do you mean I should squash them to one commit, or will you use github's squash merge to do it yourself?

**Comment by weavejester** [comment]
Unfortunately GitHub's "squash and merge" doesn't maintain a merge commit - it's more like a "squash and rebase" - so I'm afraid you need to squash the commits manually.

**Comment by luontola** [comment]
Squashed.

I moved the tests from `hiccup2.core-test` to `hiccup.compiler-test`, because they're focused on covering the edge cases in individual functions in `hiccup.compiler`.

**Review by weavejester — APPROVED** [review]
